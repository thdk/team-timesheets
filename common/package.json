{
  "name": "common",
  "version": "1.0.0",
  "description": "\"Common code for timesheets app, firebase functions, chrome extension,...\"",
  "main": "lib/index.js",
  "module": "dist/index.js",
  "jsnext:main": "dist/index.js",
  "scripts": {
    "build": "tsc",
    "build:commonjs": "tsc --module CommonJS --outDir lib",
    "build:es": "tsc"
  },
  "files": [
    "es",
    "lib"
  ],
  "author": "thdk",
  "license": "ISC",
  "dependencies": {},
  "devDependencies": {
    "typescript": "^4.4.4"
  }
}
