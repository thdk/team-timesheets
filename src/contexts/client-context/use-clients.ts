import { useConfigStore } from "../../stores/config-store/use-config-store";
import { useObserver } from "mobx-react-lite";

export const useClients = () => {
    const configStore = useConfigStore();

    return useObserver(() => ({
        clients: configStore.clients,
        addAsync: configStore.clientsCollection.addAsync.bind(configStore),
        clientId: configStore.clientId,
        get: configStore.clientsCollection.get.bind(configStore.clientsCollection),
    }));
}