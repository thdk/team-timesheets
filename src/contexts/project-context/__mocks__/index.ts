export const useProjectStore = jest.fn().mockReturnValue({
    projects: [],
    collection: {
        isFetched: true,
        docs: [],
    },
    activeProjects: [],
});
