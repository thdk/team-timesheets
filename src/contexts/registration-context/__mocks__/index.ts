export * from "./use-registration-store";
