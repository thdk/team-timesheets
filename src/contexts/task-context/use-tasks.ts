import { useConfigStore } from "../../stores/config-store/use-config-store";
import { useObserver } from "mobx-react-lite";

export const useTasks = () => {
    const configStore = useConfigStore();

    return useObserver(() => ({
        tasks: configStore.tasks,
        addAsync: configStore.tasksCollection.addAsync.bind(configStore.tasksCollection),
        taskId: configStore.taskId,
        setTaskId: configStore.setTaskId.bind(configStore),
        get: configStore.tasksCollection.get.bind(configStore.tasksCollection),
    }));
}
