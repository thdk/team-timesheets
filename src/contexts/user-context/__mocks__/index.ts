export const useUserStoreMock = {
    authenticatedUser: {
        email: "foobar@email.com",
        name: "Foobar",
        roles: {
            admin: true,
        },
    },
    divisionUser: {
        email: "foobar@email.com",
        name: "Foobar",
        roles: {
            admin: true,
        },
    },
    usersCollection: {
        isFetched: true,
    },
    divisionUsersCollection: {
        isFetched: true,
    },
    updateDivisionUser: jest.fn(),
} as ReturnType<typeof useUserStore>;

export const useUserStore = jest.fn().mockReturnValue(useUserStoreMock);
