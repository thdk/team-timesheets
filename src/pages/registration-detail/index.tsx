import * as React from 'react';
import { observer } from "mobx-react";
import { Chip, ChipSet } from '@material/react-chips';

import { Form, FormField } from '../../components/layout/form';
import { FlexGroup } from '../../components/layout/flex';
import ProjectSelect from '../../containers/projects/select';
import ClientSelect from '../../containers/clients/select';
import { withAuthentication } from '../../containers/users/with-authentication';
import { RedirectToLogin } from '../../internal';
import { TextField } from '@rmwc/textfield';
import { StoreContext } from '../../contexts/store-context';

@observer
class Registration extends React.Component {
    declare context: React.ContextType<typeof StoreContext>
    static contextType = StoreContext;

    render() {
        if (!this.context.user.authenticatedUser) {
            return <></>;
        }

        if (!this.context.timesheets.registration || !this.context.user.userId) return <></>;

        const userTasks = Array.from(this.context.user.authenticatedUser.tasks.keys());
        const { task,
            description,
            time,
            date,
            client
        } = this.context.timesheets.registration;

        const tasks = Array.from(this.context.config.tasks.docs.values())
            .filter(t => t.data && userTasks.length ? userTasks.some(userTaskId => userTaskId === t.id) : true)
            .map(t => {
                const { id: taskId, data: { name: taskName = "N/A", icon: taskIcon = undefined } = {} } = t;
                const leadingIcon = taskIcon ? <i className="material-icons mdc-chip__icon mdc-chip__icon--leading">{taskIcon}</i> : undefined;
                return (
                    <Chip leadingIcon={leadingIcon} handleSelect={this.taskClicked} id={taskId} label={taskName!} key={taskId}></Chip>
                );
            });

        return (
            <>
                <Form>
                    <FlexGroup extraCssClass="row">
                        <FormField>
                            <TextField
                                type="number"
                                outlined={true}
                                autoFocus={true}
                                tabIndex={0}
                                onChange={this.onTimeChange}
                                value={(time || "").toString()}
                                id="time"
                                label="Time"
                            />
                        </FormField>
                        <FormField first={false}>
                            <TextField
                                tabIndex={-1}
                                disabled={true}
                                value={`${date!.getFullYear()}-${date!.getMonth() + 1}-${date!.getDate()}`}
                                id="date"
                                label="Date"
                                icon="event"
                                outlined={true} />
                        </FormField>
                    </FlexGroup>
                    <FlexGroup extraCssClass="row">
                        <FormField>
                            <TextField
                                textarea
                                cols={250}
                                rows={2}
                                outlined={true}
                                tabIndex={0}
                                onChange={this.onDescriptionChange}
                                value={description}
                                id="description"
                                label="Description"
                                fullwidth
                            />
                        </FormField>
                    </FlexGroup>
                    <FlexGroup>
                        <FormField>
                            <ProjectSelect></ProjectSelect>
                        </FormField>
                        <FormField first={false}>
                            <ClientSelect
                                onChange={this.onClientChange}
                                value={client}>
                            </ClientSelect>
                        </FormField>
                    </FlexGroup>
                    <FlexGroup extraCssClass="row">
                        <FlexGroup direction="vertical">
                            <h3 className="mdc-typography--subtitle1">Select one of your standard tasks</h3>
                            <FormField>
                                <ChipSet selectedChipIds={task ? [task] : undefined} choice={true}>{tasks}</ChipSet>
                            </FormField>
                        </FlexGroup>
                    </FlexGroup>
                </Form>
            </>
        );
    }

    onDescriptionChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (this.context.timesheets.registration && this.context.timesheets.registration)
            this.context.timesheets.registration.description = event.currentTarget.value.trimLeft();
    }

    onTimeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (this.context.timesheets.registration && this.context.timesheets.registration)
            this.context.timesheets.registration.time = +(event.currentTarget.value);
    }

    taskClicked = (taskId: string, selected: boolean) => {
        if (this.context.timesheets.registration && this.context.timesheets.registration.task !== taskId && selected)
            this.context.timesheets.registration.task = taskId;
    }

    onClientChange = (value: string) => {
        if (this.context.timesheets.registration && this.context.timesheets.registration)
            this.context.timesheets.registration.client = value;
    }
}

export default withAuthentication(
    Registration,
    <RedirectToLogin />,
);
