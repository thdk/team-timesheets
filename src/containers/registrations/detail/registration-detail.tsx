import React, { useCallback } from "react";
import { observer } from "mobx-react-lite";
import { useRegistrationStore } from '../../../contexts/registration-context';
import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import { TextField } from "@rmwc/textfield";
import { ProjectSelect } from "../../projects/select/project-select";
import ClientSelect from "../../clients/select";
import { TasksChips } from "../../tasks/chips";
import { useUserStore } from "../../../contexts/user-context";
import { useTasks } from "../../../contexts/task-context";

const Tasks = observer(() => {
    const timesheets = useRegistrationStore();
    const tasksStore = useTasks();
    const user = useUserStore();

    const taskClicked = useCallback((taskId: string) => {
        if (timesheets.activeDocument && timesheets.activeDocument.task !== taskId)
            timesheets.activeDocument.task = taskId;
    }, [timesheets.activeDocument]);

    if (!user.divisionUser || !timesheets.activeDocument) {
        return null;
    }

    const userTasks = Array.from(user.divisionUser.tasks.keys());
    const tasks = tasksStore.tasks
        .filter(t => userTasks.length ? userTasks.some(userTaskId => userTaskId === t.id) : true)

    const { task } = timesheets.activeDocument;

    return tasks.length
        ? (
            <FlexGroup column>
                <h3 className="mdc-typography--subtitle1">Task</h3>
                <FormField>
                    <TasksChips
                        onTaskInteraction={taskClicked}
                        selectedTaskIds={task ? [task] : undefined}
                        choice
                        tasks={tasks}
                    />
                </FormField>
            </FlexGroup>
        )
        : null;
});

export const RegistrationDetail = observer(() => {
    const user = useUserStore();
    const timesheets = useRegistrationStore();

    const onDescriptionChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
        if (timesheets.activeDocument && timesheets.activeDocument)
            timesheets.activeDocument.description = event.currentTarget.value.trimLeft();
    }, [timesheets.activeDocument]);

    const onTimeChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
        if (timesheets.activeDocument && timesheets.activeDocument)
            timesheets.activeDocument.time = +(event.currentTarget.value);
    }, [timesheets.activeDocument]);

    const onClientChange = useCallback((value: string) => {
        if (timesheets.activeDocument)
            timesheets.activeDocument.client = value;
    }, [timesheets.activeDocument]);

    const onProjectChange = useCallback((value: string) => {
        if (timesheets.activeDocument)
            timesheets.activeDocument.project = value;
    }, [timesheets.activeDocument]);

    if (!timesheets.activeDocument || !user.divisionUser) return <></>;

    const {
        description,
        time,
        date,
        client,
        project,
    } = timesheets.activeDocument;

    return (
        <>
            <Form>
                <FlexGroup className="row">
                    <FormField>
                        <TextField
                            type="number"
                            outlined={true}
                            autoFocus={true}
                            tabIndex={0}
                            onChange={onTimeChange}
                            value={(time || "").toString()}
                            id="time"
                            label="Time"
                        />
                    </FormField>
                    <FormField first={false}>
                        <TextField
                            tabIndex={-1}
                            disabled={true}
                            value={`${date!.getFullYear()}-${date!.getMonth() + 1}-${date!.getDate()}`}
                            id="date"
                            label="Date"
                            icon="event"
                            outlined={true} />
                    </FormField>
                </FlexGroup>
                <FlexGroup className="row">
                    <FormField>
                        <TextField
                            textarea
                            cols={250}
                            rows={2}
                            outlined={true}
                            tabIndex={0}
                            onChange={onDescriptionChange}
                            value={description}
                            id="description"
                            label="Description"
                            fullwidth
                        />
                    </FormField>
                </FlexGroup>
                <FlexGroup>
                    <FormField>
                        <ProjectSelect
                            onChange={onProjectChange}
                            value={project}
                        />
                    </FormField>
                    <FormField first={false}>
                        <ClientSelect
                            onChange={onClientChange}
                            value={client}
                        />
                    </FormField>
                </FlexGroup>
                <FlexGroup className="row">
                    <Tasks />
                </FlexGroup>
            </Form>
        </>
    );
});
