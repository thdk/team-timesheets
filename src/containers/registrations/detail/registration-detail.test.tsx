import React from "react";

import { render } from "@testing-library/react";
import { RegistrationDetail } from "./registration-detail";
import { useUserStore } from "../../../contexts/user-context";
import { useRegistrationStore } from "../../../contexts/registration-context";
import { IRegistration } from "../../../../common";
import { useTasks } from "../../../contexts/task-context";
import { useClients } from "../../../contexts/client-context";
import { useProjectStore } from "../../../contexts/project-context";


jest.mock("../../../contexts/user-context", () => ({
    useUserStore: jest.fn(),
}));

jest.mock("../../../contexts/task-context");
jest.mock("../../../contexts/registration-context");
jest.mock("../../../contexts/client-context");
jest.mock("../../../contexts/project-context");

beforeEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();

    const tasks = new Map();
    tasks.set("task-1", true);
    tasks.set("task-3", true);
    (useUserStore as any).mockReturnValue({
        divisionUser: {
            id: "user-1",
            roles: {},
            tasks,
            recentProjects: [],
        },
    });
    (useClients as any).mockReturnValue({
        clients: [],
    });
    (useProjectStore as any).mockReturnValue({
        activeProjects: [],
        archivedProjects: [],
    });
});

describe("RegistrationDetail", () => {
    it("should not render when there is no authenticated user", () => {
        (useRegistrationStore as any).mockReturnValue({});
        (useUserStore as any).mockReturnValue({
            divisionUser: undefined,
        });
        const { asFragment } = render(<RegistrationDetail />);

        expect(asFragment()).toMatchSnapshot();
    });


    it("should not render when there is no registration", () => {
        (useRegistrationStore as any).mockReturnValue({});

        const { asFragment } = render(<RegistrationDetail />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render when there is a registration", () => {
        (useRegistrationStore as any).mockReturnValue({
            activeDocument: {
                date: new Date(2020, 2, 1),
                isPersisted: true,
                userId: "user-1",
                description: "Registration 1",
                task: "task-1",
            } as IRegistration,
        });

        (useTasks as any).mockReturnValue({
            tasks: [{
                id: "task-1",
                name: "Task 1",
            }],
        });

        const { asFragment } = render(<RegistrationDetail />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should only show user tasks instead of all tasks", () => {
        (useRegistrationStore as any).mockReturnValue({
            activeDocument: {
                date: new Date(2020, 2, 1),
                isPersisted: true,
                userId: "user-1",
                description: "Registration 1",
                task: "task-1",
            } as IRegistration,
        });

        (useTasks as any).mockReturnValue({
            tasks: [
                {
                    id: "task-1",
                    name: "Task 1",
                },
                {
                    id: "task-2",
                    name: "Task 2",
                },
                {
                    id: "task-3",
                    name: "Task 3",
                },
            ],
        });

        const { asFragment } = render(<RegistrationDetail />);

        expect(asFragment()).toMatchSnapshot();
    });
});
