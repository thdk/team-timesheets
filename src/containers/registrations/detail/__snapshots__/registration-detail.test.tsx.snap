// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationDetail should not render when there is no authenticated user 1`] = `<DocumentFragment />`;

exports[`RegistrationDetail should not render when there is no registration 1`] = `<DocumentFragment />`;

exports[`RegistrationDetail should only show user tasks instead of all tasks 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <div
      class="flex row horizontal"
    >
      <div
        class="first formField"
      >
        <label
          aria-labelledby="time-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--focused"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            id="time"
            tabindex="0"
            type="number"
            value=""
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="time-label"
                style="transition-duration: 0s;"
              >
                Time
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
      <div
        class="formField"
      >
        <label
          aria-labelledby="date-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--disabled mdc-text-field--with-leading-icon"
        >
          <i
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--leading"
          >
            event
          </i>
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            disabled=""
            id="date"
            tabindex="-1"
            value="2020-3-1"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="date-label"
                style="transition-duration: 0s;"
              >
                Date
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      class="flex row horizontal"
    >
      <div
        class="first formField"
      >
        <label
          aria-labelledby="description-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--textarea mdc-text-field--fullwidth mdc-text-field--outlined"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <textarea
            class="mdc-text-field__input"
            cols="250"
            id="description"
            rows="2"
            tabindex="0"
          >
            Registration 1
          </textarea>
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="description-label"
                style="transition-duration: 0s;"
              >
                Description
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      class="flex horizontal"
    >
      <div
        class="first formField"
      >
        <div
          class="mdc-select mdc-select--outlined"
          role="listbox"
        >
          <div
            class="mdc-select__anchor"
          >
            <i
              class="mdc-select__dropdown-icon"
            />
            <div
              aria-disabled="false"
              aria-haspopup="listbox"
              class="mdc-select__selected-text"
              role="button"
              tabindex="-1"
            >
               
            </div>
            <div
              class="mdc-notched-outline mdc-notched-outline--upgraded"
            >
              <div
                class="mdc-notched-outline__leading"
              />
              <div
                class="mdc-notched-outline__notch"
              >
                <span
                  class="mdc-floating-label"
                  for="select-Project"
                  style="transition-duration: 0s;"
                >
                  Project
                </span>
              </div>
              <div
                class="mdc-notched-outline__trailing"
              />
            </div>
            <select
              class="rmwc-select__native-control "
              tabindex="0"
            >
              <option
                disabled=""
                value=""
              />
              <option
                disabled=""
                value=""
              />
            </select>
          </div>
        </div>
      </div>
      <div
        class="formField"
      >
        <div
          class="mdc-select mdc-select--outlined"
          role="listbox"
        >
          <div
            class="mdc-select__anchor"
          >
            <i
              class="mdc-select__dropdown-icon"
            />
            <div
              aria-disabled="false"
              aria-haspopup="listbox"
              class="mdc-select__selected-text"
              role="button"
              tabindex="-1"
            >
               
            </div>
            <div
              class="mdc-notched-outline mdc-notched-outline--upgraded"
            >
              <div
                class="mdc-notched-outline__leading"
              />
              <div
                class="mdc-notched-outline__notch"
              >
                <span
                  class="mdc-floating-label"
                  for="client-collection"
                  style="transition-duration: 0s;"
                >
                  Client
                </span>
              </div>
              <div
                class="mdc-notched-outline__trailing"
              />
            </div>
            <select
              class="rmwc-select__native-control "
              id="client-collection"
              tabindex="0"
            >
              <option
                disabled=""
                value=""
              />
            </select>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex row horizontal"
    >
      <div
        class="flex vertical"
      >
        <h3
          class="mdc-typography--subtitle1"
        >
          Task
        </h3>
        <div
          class="first formField"
        >
          <div
            class="mdc-chip-set mdc-chip-set--choice"
          >
            <div
              class="mdc-ripple-surface mdc-chip mdc-chip--selected"
              id="task-1"
              role="row"
            >
              <div
                class="mdc-chip__ripple"
              />
              <span
                role="gridcell"
              >
                <span
                  class="mdc-chip__text__primary-action"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="mdc-chip__text"
                  >
                    Task 1
                  </span>
                </span>
              </span>
            </div>
            <div
              class="mdc-ripple-surface mdc-chip"
              id="task-3"
              role="row"
            >
              <div
                class="mdc-chip__ripple"
              />
              <span
                role="gridcell"
              >
                <span
                  class="mdc-chip__text__primary-action"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="mdc-chip__text"
                  >
                    Task 3
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RegistrationDetail should render when there is a registration 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <div
      class="flex row horizontal"
    >
      <div
        class="first formField"
      >
        <label
          aria-labelledby="time-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--focused"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            id="time"
            tabindex="0"
            type="number"
            value=""
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="time-label"
                style="transition-duration: 0s;"
              >
                Time
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
      <div
        class="formField"
      >
        <label
          aria-labelledby="date-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--disabled mdc-text-field--with-leading-icon"
        >
          <i
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--leading"
          >
            event
          </i>
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            disabled=""
            id="date"
            tabindex="-1"
            value="2020-3-1"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="date-label"
                style="transition-duration: 0s;"
              >
                Date
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      class="flex row horizontal"
    >
      <div
        class="first formField"
      >
        <label
          aria-labelledby="description-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--textarea mdc-text-field--fullwidth mdc-text-field--outlined"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <textarea
            class="mdc-text-field__input"
            cols="250"
            id="description"
            rows="2"
            tabindex="0"
          >
            Registration 1
          </textarea>
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="description-label"
                style="transition-duration: 0s;"
              >
                Description
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      class="flex horizontal"
    >
      <div
        class="first formField"
      >
        <div
          class="mdc-select mdc-select--outlined"
          role="listbox"
        >
          <div
            class="mdc-select__anchor"
          >
            <i
              class="mdc-select__dropdown-icon"
            />
            <div
              aria-disabled="false"
              aria-haspopup="listbox"
              class="mdc-select__selected-text"
              role="button"
              tabindex="-1"
            >
               
            </div>
            <div
              class="mdc-notched-outline mdc-notched-outline--upgraded"
            >
              <div
                class="mdc-notched-outline__leading"
              />
              <div
                class="mdc-notched-outline__notch"
              >
                <span
                  class="mdc-floating-label"
                  for="select-Project"
                  style="transition-duration: 0s;"
                >
                  Project
                </span>
              </div>
              <div
                class="mdc-notched-outline__trailing"
              />
            </div>
            <select
              class="rmwc-select__native-control "
              tabindex="0"
            >
              <option
                disabled=""
                value=""
              />
              <option
                disabled=""
                value=""
              />
            </select>
          </div>
        </div>
      </div>
      <div
        class="formField"
      >
        <div
          class="mdc-select mdc-select--outlined"
          role="listbox"
        >
          <div
            class="mdc-select__anchor"
          >
            <i
              class="mdc-select__dropdown-icon"
            />
            <div
              aria-disabled="false"
              aria-haspopup="listbox"
              class="mdc-select__selected-text"
              role="button"
              tabindex="-1"
            >
               
            </div>
            <div
              class="mdc-notched-outline mdc-notched-outline--upgraded"
            >
              <div
                class="mdc-notched-outline__leading"
              />
              <div
                class="mdc-notched-outline__notch"
              >
                <span
                  class="mdc-floating-label"
                  for="client-collection"
                  style="transition-duration: 0s;"
                >
                  Client
                </span>
              </div>
              <div
                class="mdc-notched-outline__trailing"
              />
            </div>
            <select
              class="rmwc-select__native-control "
              id="client-collection"
              tabindex="0"
            >
              <option
                disabled=""
                value=""
              />
            </select>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex row horizontal"
    >
      <div
        class="flex vertical"
      >
        <h3
          class="mdc-typography--subtitle1"
        >
          Task
        </h3>
        <div
          class="first formField"
        >
          <div
            class="mdc-chip-set mdc-chip-set--choice"
          >
            <div
              class="mdc-ripple-surface mdc-chip mdc-chip--selected"
              id="task-1"
              role="row"
            >
              <div
                class="mdc-chip__ripple"
              />
              <span
                role="gridcell"
              >
                <span
                  class="mdc-chip__text__primary-action"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="mdc-chip__text"
                  >
                    Task 1
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
